src/wayland/fifo/mod.rs

```diff
@@ -0,0 +1,367 @@
+                    states
+                        .cached_state
+                        .get::<FifoCachedState>()
+                        .pending()
+                        .wait_barrier = true;
+                });
+            }
+            wp_fifo_v1::Request::Destroy => {
+                if let Ok(surface) = data.upgrade() {
+                    with_states(&surface, |states| {
+                        states
+                            .data_map
+                            .get::<RefCell<FifoMarker>>()
+                            .unwrap()
+                            .borrow_mut()
+                            .0 = None;
+                    });
+                }
+            }
+            _ => unreachable!(),
+        }
+    }
+}
+
+/// State for the [`WpFifoV1`] object
+#[derive(Debug, Default, Copy, Clone)]
+pub struct FifoCachedState {
+    /// The content update requested a barrier to be set
+    pub set_barrier: bool,
+
+    /// The content update requested to wait on a previously
+    /// set barrier
+    pub wait_barrier: bool,
+}
+
+impl Cacheable for FifoCachedState {
+    fn commit(&mut self, _dh: &DisplayHandle) -> Self {
+        std::mem::take(self)
+    }
+
+    fn merge_into(self, into: &mut Self, _dh: &DisplayHandle) {
+        *into = self;
+    }
+}
+
+/// A simple barrier that can be queried and signaled
+#[derive(Debug, Clone)]
+pub struct FifoBarrier(Arc<AtomicBool>);
+
+impl PartialEq for FifoBarrier {
+    fn eq(&self, other: &Self) -> bool {
+        Arc::ptr_eq(&self.0, &other.0)
+    }
+}
+
+impl FifoBarrier {
+    /// Initialize a new barrier with the provided state
+    pub fn new(signaled: bool) -> Self {
+        Self(Arc::new(AtomicBool::new(signaled)))
+    }
+
+    /// Query the current state of the barrier
+    #[inline]
+    pub fn is_signaled(&self) -> bool {
+        self.0.load(Ordering::Relaxed)
+    }
+
+    /// Signal the barrier
+    pub fn signal(&self) {
+        self.0.store(true, Ordering::Relaxed)
+    }
+}
+
+impl Blocker for FifoBarrier {
+    fn state(&self) -> BlockerState {
+        if self.is_signaled() {
+            BlockerState::Released
+        } else {
+            BlockerState::Pending
+        }
+    }
+}
+
+/// Fifo barrier per surface state
+#[derive(Debug, Default)]
+pub struct FifoBarrierCachedState {
+    /// The barrier set for the current content update
+    pub barrier: Option<FifoBarrier>,
+}
+
+impl Cacheable for FifoBarrierCachedState {
+    fn commit(&mut self, _dh: &DisplayHandle) -> Self {
+        Self {
+            barrier: self.barrier.clone(),
+        }
+    }
+
+    fn merge_into(mut self, into: &mut Self, _dh: &DisplayHandle) {
+        let Some(barrier) = self.barrier.take() else {
+            return;
+        };
+
+        if into.barrier.as_ref() == Some(&barrier) || barrier.is_signaled() {
+            return;
+        }
+
+        if let Some(barrier) = into.barrier.replace(barrier) {
+            barrier.signal();
```

Comment:
I tried to come up with a reason why this might be problematic, but so far couldn't find one.